numba/pythonapi.py

```diff
@@ -778,76 +805,51 @@ def serialize_object(self, obj):
             self.module.__serialized[obj] = gv
         return gv
 
-    def to_native_arg(self, obj, typ):
-        if isinstance(typ, types.Record):
-            # Generate a dummy integer type that has the size of Py_buffer
-            dummy_py_buffer_type = Type.int(_helperlib.py_buffer_size * 8)
-            # Allocate the Py_buffer
-            py_buffer = cgutils.alloca_once(self.builder, dummy_py_buffer_type)
-
-            # Zero-fill the py_buffer. where the obj field in Py_buffer is NULL
-            # PyBuffer_Release has no effect.
-            zeroed_buffer = lc.Constant.null(dummy_py_buffer_type)
-            self.builder.store(zeroed_buffer, py_buffer)
-
-            buf_as_voidptr = self.builder.bitcast(py_buffer, self.voidptr)
-            ptr = self.extract_record_data(obj, buf_as_voidptr)
-
-            with cgutils.if_unlikely(self.builder,
-                                     cgutils.is_null(self.builder, ptr)):
-                self.builder.ret(ptr)
-
-            ltyp = self.context.get_value_type(typ)
-            val = self.builder.bitcast(ptr, ltyp)
-
-            def dtor():
-                self.release_record_buffer(buf_as_voidptr)
-
-        else:
-            val = self.to_native_value(obj, typ)
-
-            def dtor():
-                pass
-
-        return val, dtor
-
     def to_native_value(self, obj, typ):
         builder = self.builder
+        def c_api_error():
+            return cgutils.is_not_null(builder, self.err_occurred())
 
         if isinstance(typ, types.Object) or typ == types.pyobject:
-            return obj
+            return NativeValue(obj)
 
         elif typ == types.boolean:
             istrue = self.object_istrue(obj)
             zero = Constant.null(istrue.type)
-            return builder.icmp(lc.ICMP_NE, istrue, zero)
+            val = builder.icmp(lc.ICMP_NE, istrue, zero)
+            return NativeValue(val, is_error=c_api_error())
+
+        elif isinstance(typ, types.Integer):
+            val = self.to_native_int(obj, typ)
+            return NativeValue(val, is_error=c_api_error())
 
         elif typ in types.unsigned_domain:
```

Comment:
I think these checks for `typ` in unsigned/signed domain aren't doing anything - as far as I can see, the types that are members of these frozen sets will all be handled by the `isinstance(typ, types.Integer)` branch above.